alf/utils/data_buffer.py

```diff
@@ -287,46 +289,61 @@ def dequeue(self, env_ids=None, blocking=False):
                 self._enqueued.wait(timeout=0.2)
             return None
         else:
-            return self._dequeue(env_ids)
+            return self._dequeue(env_ids=env_ids, n=n)
 
     @atomic
-    def _dequeue(self, env_ids=None):
-        """Return earliest batch and remove it from buffer.
+    def _dequeue(self, env_ids=None, n=1):
+        """Return earliest ``n`` steps and mark them removed in the buffer.
 
         Args:
             env_ids (Tensor): If None, ``batch_size`` must be num_environments.
                 If not None, dequeue from these environments. We assume there
                 is no duplicate ids in ``env_id``. ``result[i]`` will be from
                 environment env_ids[i].
+            n (int): Number of batches to dequeue.
         Returns:
-            nested Tensors of shapes ``[batch_size, ...]``.
+            nested Tensors of shapes ``[batch_size, n, ...]``.
         Raises:
             ``AssertionError`` when not enough data is present.
         """
         with alf.device(self._device):
             env_ids = self.check_convert_env_ids(env_ids)
             current_size = self._current_size[env_ids]
             min_size = current_size.min()
-            assert min_size >= 1, (
+            assert min_size >= n, (
                 "Not all environments have enough data. The smallest data "
                 "size is: %s Try storing more data before calling dequeue" %
                 min_size)
-            pos = self._current_pos[env_ids] - current_size
-            pos = pos % self._max_length
-            indices = env_ids * self._max_length + pos  # shape [B]
-
             batch_size = env_ids.shape[0]
+            pos = self._current_pos[env_ids] - current_size  # mod done later
+            b_indices = env_ids.reshape(batch_size, 1).repeat(1, n)
+            t_range = torch.as_tensor([range(n)] * batch_size)
+            t_indices = (pos.reshape(batch_size, 1).repeat(1, n) +
```

Comment:
`pos.reshape(batch_size, 1) + t_range` is more concise and efficient because of the broadcast mechanism (https://pytorch.org/docs/stable/notes/broadcasting.html)

Reply:
interesting!